File: HackerRank Solutions/Interview Preparation Kit/Dynamic Programming/Candies.java
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.regex.*;

public class Solution {

    // Complete the candies function below.
    static long candies(int n, int[] arr) {
        int[] leftToRightRating = new int[n];
        Arrays.fill(leftToRightRating, 1);
        int[] rightToLeftRating = new int[n];
        Arrays.fill(rightToLeftRating, 1);
        for (int i = 1; i < n; i++) {
            if (arr[i] > arr[i-1]) {
                leftToRightRating[i] = leftToRightRating[i-1] + 1;
            }
        }
        for (int i = n-2; i >= 0; i--) {
            if (arr[i] > arr[i+1]) {
                rightToLeftRating[i] = rightToLeftRating[i+1] + 1;
            }
        }
        long numberOfCandies = 0L;
        for (int i = 0; i < n; i++) {
            numberOfCandies += Math.max(leftToRightRating[i], rightToLeftRating[i]);
        }
        return numberOfCandies;
    }

    private static final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) throws IOException {
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        int n = scanner.nextInt();
        scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

        int[] arr = new int[n];

        for (int i = 0; i < n; i++) {
            int arrItem = scanner.nextInt();
            scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");
            arr[i] = arrItem;
        }

        long result = candies(n, arr);

        bufferedWriter.write(String.valueOf(result));
        bufferedWriter.newLine();

        bufferedWriter.close();

        scanner.close();
    }
}

File: HackerRank Solutions/Interview Preparation Kit/Dynamic Programming/Candies.py
#!/bin/python3

import math
import os
import random
import re
import sys

# Complete the candies function below.
def candies(n, arr):
    leftToRightRatings = [1] * n
    rightToLeftRatings = [1] * n
    for i in range(1, n):
        if arr[i] > arr[i-1]:
            leftToRightRatings[i] = leftToRightRatings[i-1] + 1
    for i in range(n-2, -1, -1):
        if arr[i] > arr[i+1]:
            rightToLeftRatings[i] = rightToLeftRatings[i+1] + 1
    numberOfCandies = 0
    for i in range(n):
        numberOfCandies += max(leftToRightRatings[i], rightToLeftRatings[i])
    return numberOfCandies

if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    n = int(input())

    arr = []

    for _ in range(n):
        arr_item = int(input())
        arr.append(arr_item)

    result = candies(n, arr)

    fptr.write(str(result) + '\n')

    fptr.close()
